daemon/capabilities/manager.go

```diff
@@ -0,0 +1,120 @@
+package capabilities
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/containerd/log"
+)
+
+type Provider interface {
+	UsesSnapshotter() bool
+}
+
+// NewManager returns a new capabilities.Manager.
+// Use this instead of Manager{}.
+func NewManager() Manager {
+	cond := sync.NewCond(&sync.Mutex{})
+	return Manager{
+		stateCond: cond,
+	}
+}
+
+type Manager struct {
+	// cacheReady is  not guarded/an atomic type
+	// so that InvalidateCache is fast/doesn't have
+	// to contend with GetCapabilities.
+	cacheReady atomic.Bool
+
+	stateCond         *sync.Cond
+	refreshInProgress bool              // guarded by stateCond
+	cache             capabilitiesCache // guarded by stateCond
+}
+
+// GetCapabilities returns Capabilities according to the requested version.
+// If the requested version N <= CurrentVersion, the requested version is
+// returned. Otherwise, GetCapabilities returns the newest capabilities
+// version (CurrentVersion).
+// Capabilities are cached by the manager, so as long as the cache is valid
+// (and the cache isn't currently being refreshed) GetCapabilities should be
+// quick and have little performance impact. However, changes to daemon
+// configuration/reloads can invalidate the cache. If this happens, the first
+// call to GetCapabilities will launch a goroutine to refresh the cache.
+// Subsequent calls to GetCapabilities will block until the cache has been
+// refreshed.
+func (m *Manager) GetCapabilities(ctx context.Context, p Provider, requestedVersion int) (VersionedCapabilities, error) {
```

Comment:
Wondering; instead of passing the Provider as argument, would it make sense to pass it as option in the constructor, and keep a reference to it? i.e.;

```go
capabilities.NewManager(daemon)
```

After that, `GetCapabilities` can be called without having to pass a reference to the daemon. Doing so would probably also make sense from a perspective that the manager keeps a cache, which should be specific to the provider it used to look up things; passing a provider for each invocation would _technically_ mean I could swap it for any other provider, but the cache not taking that into consideration.

Reply:
Yeah, this sounds good to me! It makes no sense to do it any other way – otherwise, a caller is just passing a provider to be used if the cache needs to be rebuilt, but if it doesn't it could be getting cache that was built from an entirely separate provider.